ui/src/registerExperimentationService.ts

```diff
@@ -0,0 +1,112 @@
+            );
+        } catch {
+            // Best effort
+        }
+    }
+
+    ext.experimentationService = result;
+}
+
+class ExperimentationServiceAdapter implements IExperimentationServiceAdapter {
+    public wrappedExperimentationService?: tas.IExperimentationService;
+
+    public async isCachedFlightEnabled(flight: string): Promise<boolean> {
+        if (!this.wrappedExperimentationService) {
+            return false;
+        }
+
+        return this.wrappedExperimentationService.isCachedFlightEnabled(flight);
+    }
+
+    public async isLiveFlightEnabled(flight: string): Promise<boolean> {
+        if (!this.wrappedExperimentationService) {
+            return false;
+        }
+
+        return this.wrappedExperimentationService.isFlightEnabledAsync(flight);
+    }
+
+    public getCachedTreatmentVariable<T extends string | number | boolean>(name: string): T | undefined {
+        if (!this.wrappedExperimentationService) {
+            return undefined;
+        }
+
+        return this.wrappedExperimentationService.getTreatmentVariable('vscode', name);
+    }
+
+    public async getLiveTreatmentVariable<T extends string | number | boolean>(name: string): Promise<T | undefined> {
+        if (!this.wrappedExperimentationService) {
+            return undefined;
+        }
+
+        return this.wrappedExperimentationService.getTreatmentVariableAsync('vscode', name);
+    }
+}
+
+class ExperimentationTelemetry implements tas.IExperimentationTelemetry {
+    private readonly sharedProperties: { [key: string]: string } = {};
+
+    public constructor(private readonly telemetryReporter: IInternalTelemetryReporter, context: vscode.ExtensionContext) {
+        context.subscriptions.push(registerTelemetryHandler((actionContext: IActionContext) => this.handleTelemetry(actionContext)));
+    }
+
+    /**
+     * Implements `postEvent` for `IExperimentationTelemetry`.
+     * @param eventName The name of the event
+     * @param props The properties to attach to the event
+     */
+    public postEvent(eventName: string, props: Map<string, string>): void {
+        const properties: { [key: string]: string } = {};
+
+        for (const key of props.keys()) {
+            properties[key] = <string>props.get(key);
+        }
+
+        this.telemetryReporter.sendTelemetryErrorEvent(eventName, { ...properties, ...this.sharedProperties });
+    }
+
+    /**
+     * Implements `setSharedProperty` for `IExperimentationTelemetry`
+     * @param name The name of the property
+     * @param value The value of the property
+     */
+    public setSharedProperty(name: string, value: string): void {
+        this.sharedProperties[name] = value;
+    }
+
+    /**
+     * Implements a telemetry handler that adds the shared properties to the event
+     * @param actionContext The action context
+     */
+    public handleTelemetry(actionContext: IActionContext): void {
+        actionContext.telemetry.properties = {

```

Comment:
I prefer `Object.assign` so that the object reference stays the same instead of creating a new object. It's unlikely, but just in case someone is holding on to a reference to the telemetry properties